code changes for user_ldap Dynamic Group Membership

alexweirig opened the pull request:
Added new setting of “Dynamic Group Member URL”
(ldapDynamicGroupMemberURL) - see LDAP settings advanced tab.

Added public function getDynamicGroupMembers.

Updated function _groupMembers.

Updated function getUserGroups.

Commit:
code changes for user_ldap Dynamic Group Membership

Added new setting of “Dynamic Group Member URL”
(ldapDynamicGroupMemberURL) - see LDAP settings advanced tab.

Added public function getDynamicGroupMembers.

Updated function _groupMembers.

Updated function getUserGroups.

MorrisJobke commented:
cc @blizzz

LukasReschke commented:
@owncloud/ldap

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -422,11 +470,15 @@ public function getUserGroups($uid) {
-			return $groups;
 		}
 
 		//uniqueMember takes DN, memberuid the uid, so we need to distinguish
```
I am afraid this code block will run, even though groups were found by "memberOf", when $dynamicGroupMemberURL is configured. This would pull groups twice unnecessarily from LDAP. This whole method needs to be refactored. We can work around with another flag, but it won't help make it more readable and maintainable. I can help with this.

alexweirig replied:
@blizzz I think this is expected... Since I've not written the code I'm not 100% sure but, if I understand you correctly, in our case we have people being memberOf static groups, i.e. Groups where we explicitly and staticly define the membership and then the same people are members of (other!) dynamic groups through defined LDAP attributes. Therefore I think we do indeed need to first retrieve the static group memberships using the memberOf and then the dynamic group memberships using the filters.
I will have a look at the full code tomorrow maybe I did a bad coy/paste of the changed code.

alexweirig replied:
@blizzz I checked the code and the copy/paste was correct. Just to make sure we get it right, here's who the processing should be:
1) retrieve the user groups based on group membership as it had been in owncloud until now
2) if dynamicGroupMemberURL is empty return the list of groups found
3) otherwise
4) retrieve the user groups based on dynamic group processing
5) merge the result from step 1 and step 4
6) return the merged list of groups found

blizzz replied:
@alexweirig true, but with 3) comes an issue: the standard group memberships as read in 1) are read again in line 504f.

alexweirig replied:
@blizzz Ahh OK I see the problem.
That's because that part of the code is the "fallback" if the memberOf was not returning groups and now we're kind of faking this by saying that we need to continue because we have to process the dynamic groups.

Actually I'm wondering if we should process the dynamic group membership first, store that result and before returning eventually do a merge of the two results? 
Would that be possible?

blizzz replied:
I think yes, this part can be moved after line 454, where $dynamicGroupMemberURL is defined. Then the $groups array is filled with the dynamic groups, and it won't be overwritten, other groups are just added.

alexweirig replied:
@blizzz Thanks, I think that should be not too complicated. what should I do? Should I edit the file here in the PR and move the lines up in the code?

blizzz replied:
Yes, you can try to move the whole block starting line 513 up. The changes in lines 473-450 are to be reverted, subsequently.

Either you do it directly here in Github (whether this is so  convenient…), or you do it where you have the code locally, do a ``commit -m "move determining dynamic group memberships up"  apps/user_ldap/group_ldap.php
`` and just push per ``git push``. This should then be updated automatically.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -458,6 +510,39 @@ public function getUserGroups($uid) {
+			// look through dynamic groups to add them to the result array if needed
+			$groupsToMatch = $this->access->fetchListOfGroups(
+				$this->access->connection->ldapGroupFilter,array('dn',$dynamicGroupMemberURL));
+			foreach($groupsToMatch as $value) {
```
minor: variable names (here $value) short be self-descriptive. $memberUrl in this case. I don't consider this a blocker however.

alexweirig replied:
Agreed

blizzz commented:
@alexweirig this looks good, except my concern at https://github.com/owncloud/core/pull/21838#discussion_r50887904, were I offered my help. I'd confirm and if so refactor the method tomorrow.

Commit:
Refactored code to avoid performance problem

Moved the dynamic group processing to the top and removed condition in memberOf processing.

Also, changed variable name $value to $memberUrl

alexweirig commented:
@blizzz I've committed the changes.
I moved the code to the top and also renamed that $value variable into $memberUrl.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -405,6 +451,41 @@ public function getUserGroups($uid) {
+					}
+				} else {
+					\OCP\Util::writeLog('user_ldap', 'No search filter found on member url '.
+						'of group ' . $dnGroup, \OCP\Util::DEBUG);
```
$dnGroup is undefined. I assume should be here instead? $memberUrl[$dynamicGroupMemberURL][0]

blizzz commented:
Except of the undefined variable :+1: 
When you fix this, I will open another branch so that unit tests are running.
Integration tests succeed.

@owncloud/ldap @MorrisJobke one more reviewer por favor?

Commit:
Fixed undefined variable $dnGroup and variable name

I renamed $memberURL into $dynamicGroup and print that variable in the writeLog.

I think this makes more sense.

alexweirig commented:
@blizzz I've changed the code to fix the undefined variable issue.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -405,6 +451,41 @@ public function getUserGroups($uid) {
+					}
+				} else {
+					\OCP\Util::writeLog('user_ldap', 'No search filter found on member url '.
+						'of group ' . $dynamicGroup, \OCP\Util::DEBUG);
```
Now the variable is defined, but it because it is an array an output would just like like ``No search filter found on member url of group Array``. The easiest solution to this is use ``print_r($dynamicGroup, true)``, i.e.

```php
\OCP\Util::writeLog('user_ldap', 'No search filter found on member url '.
						'of group ' . print_r($dynamicGroup, true), \OCP\Util::DEBUG);
```

(The alternative would be to get the modern Logger from \oc::$server, which is more powerful, but properly done it would require to change class signatures in many places for proper dependency injection. I don't want to bother you with this, seems overkill.)

blizzz commented:
Also, can you do a rebase against ownCloud master? It's some git magic. If not, I'll do it when creating a branch to run unit tests.

Commit:
change error logging

replaced variable with print_r call

alexweirig commented:
@blizzz I've changed the code to include the print_r.

I wanted to rebase but, from what I read, that's not possible in the web interface??

MorrisJobke commented:
> I wanted to rebase but, from what I read, that's not possible in the web interface??

Yes, because there is a merge conflict.

alexweirig commented:
@MorrisJobke OK? and the interface says:
This branch has conflicts that must be resolved
Only those with write access to this repository can merge pull requests.

Not sure if I can help

blizzz commented:
I'll submit a branch for  our unit tests and will rebase there, so we can merge that one. No problems.

blizzz commented:
Here it goes: https://github.com/owncloud/core/pull/22025

DeepDiver1975 commented:
> Here it goes: #22025

merged
